sideline-eglot.el

```diff
@@ -57,12 +63,28 @@
 
 (defun sideline-eglot--async-candidates (callback &rest _)
   "Request eglot's candidates."
-  (if sideline-eglot--ht-candidates
-      (ht-clear sideline-eglot--ht-candidates)
-    (setq sideline-eglot--ht-candidates (ht-create)))
-  (dolist (row (ignore-errors (eglot-code-actions (point))))
-    (ht-set sideline-eglot--ht-candidates (cl-getf row :title) row))
-  (funcall callback (ht-keys sideline-eglot--ht-candidates)))
+  (jsonrpc-async-request
+   (eglot-current-server)
+   :textDocument/codeAction
+   (list :textDocument (eglot--TextDocumentIdentifier)
+         :range (list :start (eglot--pos-to-lsp-position (point))
+                      :end (eglot--pos-to-lsp-position nil))
+         :context
+         `(:diagnostics
+           [,@(cl-loop for diag in (flymake-diagnostics (point) nil)
+                       when (cdr (assoc 'eglot-lsp-diag
+                                        (eglot--diag-data diag)))
+                       collect it)]))
+   :success-fn
+   (lambda (resp)
+     (let ((actions (append resp nil)))
+       (if sideline-eglot--ht-candidates
+           (ht-clear sideline-eglot--ht-candidates)
+         (setq sideline-eglot--ht-candidates (ht-create)))
+       (dolist (row actions)
+         (ht-set sideline-eglot--ht-candidates (cl-getf row :title) row))
+       (funcall callback (ht-keys sideline-eglot--ht-candidates))))
```

Comment:
Apparently this lambda doesn't detect callback arg, any suggestion?

Reply:
Sorry, I'm not sure if I get it. What do you mean? 🤔

Reply:
```
(defun sideline-eglot--async-candidates (callback &rest _)
```
I mean `callback` argument in this function ^^

```
  "Request eglot's candidates."
  (jsonrpc-async-request
   (eglot-current-server)
   :textDocument/codeAction
   (list :textDocument (eglot--TextDocumentIdentifier)
         :range (list :start (eglot--pos-to-lsp-position (point))
                      :end (eglot--pos-to-lsp-position nil))
         :context
         `(:diagnostics
           [,@(cl-loop for diag in (flymake-diagnostics (point) nil)
                       when (cdr (assoc 'eglot-lsp-diag
                                        (eglot--diag-data diag)))
                       collect it)]))
   :success-fn
   (lambda (resp)
     (let ((actions (append resp nil)))
       (if sideline-eglot--ht-candidates
           (ht-clear sideline-eglot--ht-candidates)
         (setq sideline-eglot--ht-candidates (ht-create)))
       (dolist (row actions)
         (ht-set sideline-eglot--ht-candidates (cl-getf row :title) row))
       (funcall callback (ht-keys sideline-eglot--ht-candidates)))

```
                     ^^
In this **funcall** `callback` is not "detected" and throw error that is not defined.

Reply:
What error did you get? 🤔 You should be able to display sidelines in any circumstances. 

Other async implementation may help a little:

- https://github.com/emacs-sideline/sideline-flycheck/blob/master/sideline-flycheck.el
- https://github.com/emacs-sideline/sideline-lsp/blob/master/sideline-lsp.el

Reply:
I think it was a problem with my emacs session, now it works.

Reply:
In sideline-flycheck, they used a global variable to track down the callback. Maybe that's one way to approach it?

Reply:
> In sideline-flycheck, they used a global variable to track down the callback. Maybe that's one way to approach it?

Thank you, now it works!